### PR TITLE
[SYCL] Correctly handle missing spec constants in set_specialization_…

### PR DESCRIPTION
**Pull request by alexbatashev**
…constant

If neither of device images uses specialization constant, store it in a temporary buffer.
Also fix sycl::join behavior:
- If two or more kernel_bundles are joined, the resulting bundle will
  have all of the specialization constants set in either of bundles.
- If two or more kernel_bundles have spec constant value set, the
  resulting kernel_bundle also has this spec constant set, but the value
  is undefined.

### CODE REVIEW

**Comment by alexbatashev** [comment]
@intel/llvm-reviewers-runtime  friendly ping

File: sycl/source/detail/kernel_bundle_impl.hpp [hunk 1]
```diff
@@ -260,7 +260,43 @@ class kernel_bundle_impl {
               LessByHash<device_image_plain>{});
     const auto DevImgIt =
         std::unique(MDeviceImages.begin(), MDeviceImages.end());
+
+    // Save device images to be removed to extract their spec constant values.
+    std::vector<device_image_plain> RemovedImages{DevImgIt,
```
Comment (s-kanaev) [inline sycl/source/detail/kernel_bundle_impl.hpp:265]: This variable isn't used.

Comment (alexbatashev) [reply sycl/source/detail/kernel_bundle_impl.hpp:265]: Nice catch, thanks

**Review by romanovvlad — DISMISSED** [review]

File: sycl/source/detail/kernel_bundle_impl.hpp [hunk 2]
```diff
@@ -261,35 +261,33 @@ class kernel_bundle_impl {
     const auto DevImgIt =
```
Comment (romanovvlad) [inline sycl/source/detail/kernel_bundle_impl.hpp:261]: NIT, Consider moving the loop before `unique` and `sort` to make it more easy to understand that `erase` happens as part of `sort + unique + erase` sequence.

**Review by romanovvlad — APPROVED** [review]

File: sycl/source/detail/kernel_bundle_impl.hpp (hunk 2 above)
Comment (alexbatashev) [reply sycl/source/detail/kernel_bundle_impl.hpp:261]: I'll submit it as a separate patch

**Comment by bader** [comment]
@alexbatashev, this PR broke post-commit.

/home/runner/work/llvm/llvm/src/sycl/source/detail/kernel_bundle_impl.hpp:275:19: error: loop variable 'SpecConst' of type 'const std::pair<std::string, std::vector<device_image_impl::SpecConstDescT>> &' (aka 'const pair<basic_string<char>, vector<device_image_impl::SpecConstDescT>> &') binds to a temporary constructed from type 'std::_Rb_tree_const_iterator<std::pair<const std::basic_string<char>, std::vector<sycl::detail::device_image_impl::SpecConstDescT>>>::reference' (aka 'const std::pair<const std::basic_string<char>, std::vector<sycl::detail::device_image_impl::SpecConstDescT>> &') [-Werror,-Wrange-loop-construct]
                 &SpecConst : SpecConsts) {
                  ^

Please, fix ASAP.

**Pull request merged**